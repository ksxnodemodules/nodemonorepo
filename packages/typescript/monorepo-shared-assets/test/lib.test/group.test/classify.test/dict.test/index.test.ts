import {iter, group} from '../../../../../index'
const classify = group.classify.dict
const getList = () => iter.fns.range(16)
const dividers = Array.from(iter.fns.range(2).up.to(16))

class ValueCarrier {
  readonly value: number

  constructor (value: number) {
    this.value = value
  }

  equal (cmp: ValueCarrier) {
    return this.value === cmp.value
  }
}

const repeatedList = Array
  .from(iter.fns.repeat(getList(), 3))
  .map(x => new ValueCarrier(x))

const duplicationChecker = (a: ValueCarrier, b: ValueCarrier) => a.equal(b)

describe('single distribution', () => {
  it('in basic use', () => {
    const classifier = (x: number) => `Ring(4, ${x % 4})`
    expect(classify.singleDistribute(getList(), classifier)).toMatchSnapshot()
  })

  it('with specified duplicationChecker', () => {
    const classifier = (x: ValueCarrier) => `Ring(4, ${x.value % 4})`
    expect(classify.singleDistribute(repeatedList, classifier, duplicationChecker)).toMatchSnapshot()
  })
})

describe('multiple distribution', () => {
  it('in basic use', () => {
    const classifier = (x: number) => dividers
      .filter(y => ![0, 1, x].includes(y) && !(x % y))
      .map(y => `x${y}`)

    expect(classify.multiDistribute(getList(), classifier)).toMatchSnapshot()
  })

  it('with specified duplicationChecker', () => {
    const classifier = (x: ValueCarrier) => dividers
      .filter(y => ![0, 1, x.value].includes(y) && !(x.value % y))
      .map(y => `x${y}`)

    expect(classify.multiDistribute(repeatedList, classifier, duplicationChecker)).toMatchSnapshot()
  })
})
