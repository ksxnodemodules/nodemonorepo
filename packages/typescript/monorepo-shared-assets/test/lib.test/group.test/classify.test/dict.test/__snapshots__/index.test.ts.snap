// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple distribution in basic use 1`] = `
Object {
  "classified": Object {
    "x10": Array [
      0,
    ],
    "x11": Array [
      0,
    ],
    "x12": Array [
      0,
    ],
    "x13": Array [
      0,
    ],
    "x14": Array [
      0,
    ],
    "x15": Array [
      0,
    ],
    "x2": Array [
      0,
      4,
      6,
      8,
      10,
      12,
      14,
    ],
    "x3": Array [
      0,
      6,
      9,
      12,
      15,
    ],
    "x4": Array [
      0,
      8,
      12,
    ],
    "x5": Array [
      0,
      10,
      15,
    ],
    "x6": Array [
      0,
      12,
    ],
    "x7": Array [
      0,
      14,
    ],
    "x8": Array [
      0,
    ],
    "x9": Array [
      0,
    ],
  },
  "unclassified": Array [
    1,
    2,
    3,
    5,
    7,
    11,
    13,
  ],
}
`;

exports[`multiple distribution with specified duplicationChecker 1`] = `
Object {
  "classified": Object {
    "x10": Array [
      ValueCarrier {
        "value": 0,
      },
    ],
    "x11": Array [
      ValueCarrier {
        "value": 0,
      },
    ],
    "x12": Array [
      ValueCarrier {
        "value": 0,
      },
    ],
    "x13": Array [
      ValueCarrier {
        "value": 0,
      },
    ],
    "x14": Array [
      ValueCarrier {
        "value": 0,
      },
    ],
    "x15": Array [
      ValueCarrier {
        "value": 0,
      },
    ],
    "x2": Array [
      ValueCarrier {
        "value": 0,
      },
      ValueCarrier {
        "value": 4,
      },
      ValueCarrier {
        "value": 6,
      },
      ValueCarrier {
        "value": 8,
      },
      ValueCarrier {
        "value": 10,
      },
      ValueCarrier {
        "value": 12,
      },
      ValueCarrier {
        "value": 14,
      },
    ],
    "x3": Array [
      ValueCarrier {
        "value": 0,
      },
      ValueCarrier {
        "value": 6,
      },
      ValueCarrier {
        "value": 9,
      },
      ValueCarrier {
        "value": 12,
      },
      ValueCarrier {
        "value": 15,
      },
    ],
    "x4": Array [
      ValueCarrier {
        "value": 0,
      },
      ValueCarrier {
        "value": 8,
      },
      ValueCarrier {
        "value": 12,
      },
    ],
    "x5": Array [
      ValueCarrier {
        "value": 0,
      },
      ValueCarrier {
        "value": 10,
      },
      ValueCarrier {
        "value": 15,
      },
    ],
    "x6": Array [
      ValueCarrier {
        "value": 0,
      },
      ValueCarrier {
        "value": 12,
      },
    ],
    "x7": Array [
      ValueCarrier {
        "value": 0,
      },
      ValueCarrier {
        "value": 14,
      },
    ],
    "x8": Array [
      ValueCarrier {
        "value": 0,
      },
    ],
    "x9": Array [
      ValueCarrier {
        "value": 0,
      },
    ],
  },
  "unclassified": Array [
    ValueCarrier {
      "value": 1,
    },
    ValueCarrier {
      "value": 2,
    },
    ValueCarrier {
      "value": 3,
    },
    ValueCarrier {
      "value": 5,
    },
    ValueCarrier {
      "value": 7,
    },
    ValueCarrier {
      "value": 11,
    },
    ValueCarrier {
      "value": 13,
    },
    ValueCarrier {
      "value": 1,
    },
    ValueCarrier {
      "value": 2,
    },
    ValueCarrier {
      "value": 3,
    },
    ValueCarrier {
      "value": 5,
    },
    ValueCarrier {
      "value": 7,
    },
    ValueCarrier {
      "value": 11,
    },
    ValueCarrier {
      "value": 13,
    },
    ValueCarrier {
      "value": 1,
    },
    ValueCarrier {
      "value": 2,
    },
    ValueCarrier {
      "value": 3,
    },
    ValueCarrier {
      "value": 5,
    },
    ValueCarrier {
      "value": 7,
    },
    ValueCarrier {
      "value": 11,
    },
    ValueCarrier {
      "value": 13,
    },
  ],
}
`;

exports[`single distribution in basic use 1`] = `
Object {
  "Ring(4, 0)": Array [
    0,
    4,
    8,
    12,
  ],
  "Ring(4, 1)": Array [
    1,
    5,
    9,
    13,
  ],
  "Ring(4, 2)": Array [
    2,
    6,
    10,
    14,
  ],
  "Ring(4, 3)": Array [
    3,
    7,
    11,
    15,
  ],
}
`;

exports[`single distribution with specified duplicationChecker 1`] = `
Object {
  "Ring(4, 0)": Array [
    ValueCarrier {
      "value": 0,
    },
    ValueCarrier {
      "value": 4,
    },
    ValueCarrier {
      "value": 8,
    },
    ValueCarrier {
      "value": 12,
    },
  ],
  "Ring(4, 1)": Array [
    ValueCarrier {
      "value": 1,
    },
    ValueCarrier {
      "value": 5,
    },
    ValueCarrier {
      "value": 9,
    },
    ValueCarrier {
      "value": 13,
    },
  ],
  "Ring(4, 2)": Array [
    ValueCarrier {
      "value": 2,
    },
    ValueCarrier {
      "value": 6,
    },
    ValueCarrier {
      "value": 10,
    },
    ValueCarrier {
      "value": 14,
    },
  ],
  "Ring(4, 3)": Array [
    ValueCarrier {
      "value": 3,
    },
    ValueCarrier {
      "value": 7,
    },
    ValueCarrier {
      "value": 11,
    },
    ValueCarrier {
      "value": 15,
    },
  ],
}
`;
